### PR TITLE
Wrappers

### PR DESCRIPTION
**Pull request by farizrahman4u**
Wrappers for Highway and time distributed layers. Example:

```python
from keras.layers.wrappers import *
model=Sequential()
model.add(TimeDistributed(Highway(Dense(input_dim=10, output_dim=10))))
model.add(TimeDistributed(Highway(Dense(output_dim=10))))
model.compile(loss='mse',optimizer='sgd')
```

### CODE REVIEW

File: keras/layers/wrapers.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+        output_dim = self.output_shape[1]
+        assert input_dim ==  output_dim, 'Input and output shapes should be equal for highway layers.'
+        self.input_dim = input_dim
+        self.W = self.init((input_dim, input_dim))
+        self.b = K.variable(np.ones((input_dim,)) * self.transform_bias)
+        self.params += [self.W, self.b]
+
+    def get_output(self, train=False):
+        X = self.get_input(train)
+        alpha = K.sigmoid(K.dot(X, self.W) + self.b)
```
Comment (lukedeo) [inline keras/layers/wrapers.py]: I think the transform/carry weight matrix/bias should be separate from `self.W` and `self.b`.

**Comment by fchollet** [comment]
This is cool. But one major issue: both wrappers are hard-coded to only work for 2D inputs. But the point of having wrappers for TimeDistributed (and Highway) is to be able to run them on arbitrary layers, and specifically convolutional layers (to create convolutional highways, which is the main application of highway networks, and to create convolutional RNNs, which is really missing in Keras).

It wouldn't be too difficult to use the `input_shape` of each wrapper to adapt the contents to a 4D input. For TimeDistributed, some minor modifications may be required elsewhere to make Keras accept 5D inputs.

**Comment by oeway** [comment]
Hi,

I am trying to make a TimeDistributed wrapper for arbitrary layer, it works for the Dense layer, but  for layers such as Convolution2D it's not working. It can be compiled, but if I try to predict, it complaining that  "...ValueError: Input dimension mis-match....". 
Can anyone help me out? 

The goal is to build a recurrent convolutional network (see #1558).

``` python
class TimeDistributedXD(MaskedLayer, Wrapper):

    def __init__(self, layer, time_steps=None, **kwargs):
        self.time_steps = time_steps
        if hasattr(layer, 'input_ndim'):
            self.input_ndim = layer.input_ndim+1
        super(TimeDistributedND, self).__init__(layer, **kwargs)

    def get_output_shape(self):
        input_shape = self.input_shape
        return (input_shape[0], input_shape[1],) + self.layer.output_shape[1:]

    def get_input_shape(self):
        timesteps = self.time_steps
        layer_input_shape =self.layer.input_shape
        ret = (layer_input_shape[0],timesteps,)+layer_input_shape[1:]
        return  ret

    def build(self):
        input_shape = self.input_shape
        layer_input_shape = (input_shape[0],)+input_shape[2:]
        self.layer.set_input_shape(layer_input_shape)
        self.set_params()

    def get_output(self, train=False):
        X = self.get_input(train)
        def step(x, states):
            self.t_ = x
            return self.layer(x), []

        return K.rnn(step, X, [])[1]

```

**Comment by jeammimi** [comment]
Your error was due to a bug in the rnn backend from theano  [#1565] (https://github.com/fchollet/keras/pull/1565) which was not dealing properly with 5D
inputs

**Comment by jocicmarko** [comment]
What's the progress with this PR? TimeDistributed and Highway wrappers will be very useful.

**Comment by farizrahman4u** [comment]
Am on it. Its actually functional. Needs more work though.

**Comment by oeway** [comment]
+1
Cool, I am testing the TimeDistributed wrapper with Convolution2D layers and Maxpooling2D layers as well. The version I posted previously is actually working quite well, with a time distributed CNN mixed with some LSTM-Convolution layers. The compiling and training is very slow though.

**Comment by fchollet** [comment]
> Am on it. Its actually functional. Needs more work though.

Please publish it as a PR even if it's not ready, so that it is open to discussion and code review. This is a pretty important feature and it would benefit from transparency.

**Comment by farizrahman4u** [comment]
@fchollet 

This is a PR.

**Comment by farizrahman4u** [comment]
The sample code given in my first comment works as of now. Only few tweaks would be required to get the timedistributed wrapper work with nD layers (especially conv2d).

**Comment by eyaler** [comment]
Hi,

I made some changes to wrappers.py/TimeDistributed to make with work with Convolution2D:

`    def get_output_shape(self):
        input_shape = self.input_shape
        return tuple([input_shape[0], input_shape[1]] + list(self.layer.output_shape[1:]))

    def get_input_shape(self):
        layer_input_shape = self.layer.input_shape
        self.input_ndim = 1 + len(layer_input_shape)
        timesteps = None if not hasattr(self, '_input_shape') else self._input_shape[1]
        return  tuple([layer_input_shape[0], timesteps] + list(layer_input_shape[1:]))

    def build(self):
        input_shape = self.input_shape
        self.layer.set_input_shape(tuple([input_shape[0]] + list(input_shape[2:])))
        self.set_params()`

Also I applied https://github.com/fchollet/keras/pull/1565

**Comment by eyaler** [comment]
I made some changes to wrappers.py/TimeDistributed to make it work with Convolution2D

`def get_output_shape(self):
        input_shape = self.input_shape
        return tuple([input_shape[0], input_shape[1]] + list(self.layer.output_shape[1:]))

    def get_input_shape(self):
        layer_input_shape = self.layer.input_shape
        self.input_ndim = 1 + len(layer_input_shape)
        timesteps = None if not hasattr(self, '_input_shape') else self._input_shape[1]
        return  tuple([layer_input_shape[0], timesteps] + list(layer_input_shape[1:]))

    def build(self):
        input_shape = self.input_shape
        self.layer.set_input_shape(tuple([input_shape[0]] + list(input_shape[2:])))
        self.set_params()`

Also I applied https://github.com/fchollet/keras/pull/1565

**Comment by farizrahman4u** [comment]
@eyaler Thanks!

**Comment by farizrahman4u** [comment]
@eyaler  One question: Is #1565 necessary for TimeDistributed Conv2D ?

**Comment by eyaler** [comment]
yup. not sure why it is closed without merge

**Comment by farizrahman4u** [comment]
@fchollet Why wan't #1565 merged? Its still hard coded for 3D tensor around that area.

**Comment by fchollet** [comment]
@farizrahman4u I believe I intended to merge it, but it got further down the stack as other things piled up. You can reopen the same PR, or make it part of the current PR.

**Comment by fchollet** [comment]
> There should be 2 different step functions.. one for with mask and one for without.

Is there a measurable performance difference? If so, then yes, there should be 2 step functions (which is already the case in the TF implementation).

**Comment by AntreasAntoniou** [comment]
This feature is super useful. I am currently using it for a Time Distributed CNN. Is there an example I could refer to, in order to build it properly?

**Comment by jarfo** [comment]
A suggestion for the TimeDistributed wrapper is to temporally collapse batch and time dimensions, then perform a single layer call. The following code works on Theano

    def get_output(self, train=False):
        X = self.get_input(train)
        X_shape = K.shape(X)
        x_shape = (X_shape[0]*X_shape[1], ) + tuple(X_shape[i] for i in range(2, K.ndim(X)))
        x = K.reshape(X, x_shape)
        y = self.layer(x)
        y_shape = K.shape(y)
        Y_shape = (X_shape[0], X_shape[1]) + tuple(y_shape[i] for i in range(1, K.ndim(y)))
        Y = K.reshape(y, Y_shape)
        return Y

The tensorflow backend seems to need 'int' values in the reshape shape. One workaround in this case is to map to int the shape values, for example, X_shape = map(int, K.shape(X)).
This int conversion can be performed in the shape or reshape functions of the tensorflow backend. All the collapse and expand code can also be moved to specific reshape functions in the backend in both cases.

**Comment by eyaler** [comment]
i also needed to delete this line:
    input_ndim = 3

**Comment by lemuriandezapada** [comment]
You also need to change this line
self.reverse.set_weights(weights[:nw/2])  to self.reverse.set_weights(weights[nw/2:])
And replace all nw/2 to int(nw/2) since at least in python 3 it gets casted back down to floats

And implement a non_trainable_weights property

**Comment by barvinograd** [comment]
+1 for the concept

**Comment by farizrahman4u** [comment]
TODO : 
Residual wrapper
Tests

**Comment by eyaler** [comment]
if this is still relevant, another patch by @barvinograd  to allow nested time-distributed models:

def build(self):
       input_shape = self.input_shape
       if not hasattr(self.layer, 'input'):
           # already has input if lasyer is container
           self.layer.set_input_shape(tuple([input_shape[0]] + list(input_shape[2:])))
       self.set_params()

**Comment by farizrahman4u** [comment]
Moved to #1828

**Pull request closed without merging**